tests/tests/filter.cpp

```diff
@@ -0,0 +1,532 @@
+                             decltype(fn)>;
+  }([](int const &) -> Error { throw 0; })); // allow binding const lvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t &, decltype(fn),
+                             decltype(onError)>;
+  }([](int &) -> bool { throw 0; })); // disallow binding lvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t &, decltype(truePred),
+                             decltype(fn)>;
+  }([](int &) -> Error { throw 0; })); // disallow binding lvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t const &, decltype(fn),
+                             decltype(onError)>;
+  }([](int &) -> bool { throw 0; })); // disallow removing const
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t const &, decltype(truePred),
+                             decltype(fn)>;
+  }([](int &) -> Error { throw 0; })); // disallow removing const
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([](int &) -> bool { throw 0; })); // disallow lvalue binding to rvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([](int &) -> Error { throw 0; })); // disallow lvalue binding to rvalue
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([](std::string) -> bool { throw 0; })); // wrong type
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([](std::string) -> Error { throw 0; })); // wrong type
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([]() -> bool { throw 0; })); // wrong arity
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([]() -> Error { throw 0; })); // wrong arity
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(fn),
+                             decltype(onError)>;
+  }([](int, int) -> bool { throw 0; })); // wrong arity
+
+  static_assert(not [](auto &&fn) constexpr {
+    return monadic_invocable<filter_t, operand_t, decltype(truePred),
+                             decltype(fn)>;
+  }([](int, int) -> Error { throw 0; })); // wrong arity
+
+  // rvalue operand
+  // --------------
+
+  // TODO: fix these two if needed:
```

Comment:
Do we actually need these? I'm probably just not seeing the use-case here. Please explain how a user of this library can get this unless they spell it out in their lambda? And if they do.. why would we want this to be allowed? :goberserk:

Reply:
It's fine. For predicator we cannot use an rvalue-ref, even if its const, because we can only give it a const lvalue (and binding an lvalue to rvalue-ref is forbidden). For error generator, it just works. I added tests.